crates/containerd-shim-wasm/src/sandbox/containerd.rs

```diff
@@ -137,23 +431,144 @@ impl Client {
+
+                // The original image is considered a root object, by adding a ref to the new compiled content
+                // We tell to containerd to not garbage collect the new content until this image is removed from the system
+                // this ensures that we keep the content around after the lease is dropped
+                log::debug!("updating content with precompile digest to avoid garbage collection");
+                let mut image_content = self.get_info(image_digest.clone())?;
+                image_content.labels.insert(
+                    "containerd.io/gc.ref.content.precompile".to_string(),
+                    precompiled_content.digest.clone(),
+                );
+                self.update_info(image_content)?;
+
+                Ok((
+                    vec![WasmLayer {
+                        config: image_config_descriptor.clone(),
+                        layer: precompiled,
+                        precompiled: true,
+                    }],
+                    platform,
+                ))
+            }
+            _ => {
+                log::info!("using module from OCI layers");
+                let layers = manifest
+                    .layers()
+                    .iter()
+                    .filter(|x| is_wasm_layer(x.media_type(), T::supported_layers_types()))
+                    .map(|config| {
+                        self.read_content(config.digest()).map(|module| WasmLayer {
+                            config: config.clone(),
+                            layer: module,
+                            precompiled: false,
+                        })
+                    })
+                    .collect::<Result<Vec<_>>>()?;
+                Ok((layers, platform))
+            }
+        }
     }
 }
 
 fn is_wasm_layer(media_type: &MediaType, supported_layer_types: &[&str]) -> bool {
     supported_layer_types.contains(&media_type.to_string().as_str())
 }
+
+#[cfg(test)]
+mod tests {
+    use std::path::PathBuf;
+
+    use super::*;
+
+    #[test]
+    fn test_save_content() {
+        let path = PathBuf::from("/run/containerd/containerd.sock");
+        let path = path.to_str().unwrap();
+        let client = Client::connect(path, "test-ns").unwrap();
+        let data = b"hello world".to_vec();
+
+        let expected = digest(data.clone());
+        let expected = format!("sha256:{}", expected);
+
+        let returned = client
+            .save_content(
+                data,
+                "original".to_string(),
+                &RuntimeInfo {
+                    name: "test",
+                    version: "0.0.0",
+                },
+            )
+            .unwrap();
+        assert_eq!(expected, returned.digest.clone());
+
+        let data = client.read_content(returned.digest.clone()).unwrap();
+        assert_eq!(data, b"hello world");
+
+        // need to drop the lease to be able to create a second one
+        drop(returned);
```

Comment:
I'd like to see a test case where the `drop()` function is not invoked explictly. What will happen?

Reply:
you will get an error, I can add a test case

Reply:
added an assert